### PR TITLE
ifacestate: only rewrite security profiles if the system-key changes

### PR DESCRIPTION
**Pull request by mvo5**
We currently unconditionally rewrite security profiles on each
restart of snapd. However this is not really needed, we only need
to rewrite them if snapd itself changes or if security properties
(like apparmor support in the kernel) changes. This PR addresses
this.

### CODE REVIEW

File: overlord/ifacestate/helpers.go [hunk 1]
```diff
@@ -156,7 +174,8 @@ func (m *InterfaceManager) regenerateAllSecurityProfiles() error {
 		}
 	}
 
-	return nil
+	sk := interfaces.SystemKey()
+	return osutil.AtomicWriteFile(dirs.SnapSystemKeyFile, []byte(sk), 0644, 0)
```
Comment (zyga) [inline overlord/ifacestate/helpers.go]: Can you please add a comment that explains why the system key is in a separate file.

File: interfaces/system_key.go [hunk 2]
```diff
@@ -35,32 +35,53 @@ type systemKey struct {
...
 
 	// Add apparmor-feature, note that ioutil.ReadDir() is already sorted.
 	//
 	// We prefix the dirs.GlobalRootDir (which is usually "/") to make
 	// this testable.
-	mySystemKey.AppArmorFeatures = release.AppArmorFeatures()
+	sk.AppArmorFeatures = release.AppArmorFeatures()
 
-	return &mySystemKey
+	return &sk
```
Comment (zyga) [inline interfaces/system_key.go]: Nitpick, you could shorten this to
```
buildID, err := osutil.MyBuildID()
if err != nil {
    buildID = ""
}
return &systemKey{BuildID: buildID, AppArmorFeatures: release.AppArmorFeatures()}
```

File: overlord/ifacestate/helpers.go [hunk 3]
```diff
@@ -108,14 +114,26 @@ func (m *InterfaceManager) addSnaps() error {
...
+	onDiskSystemKey, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
+	if os.IsNotExist(err) {
+		return true
+	}
+	if err != nil {
+		logger.Noticef("cannot read system-key file: %s", err)
+		return true
+	}
+
+	return string(onDiskSystemKey) != currentSystemKey
```
Comment (zyga) [inline overlord/ifacestate/helpers.go]: This is so awesome! Thank you @mvo5 :)

File: tests/main/snap-system-key/task.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        systemctl start snapd.service
+        while [ "$(systemctl show -pActiveState snapd.service)" != "ActiveState=active" ]; do
+            systemctl show -pActiveState snapd.service
+            sleep 1
+        done
+    }
+
+    echo "Ensure a valid system-key file is on-disk"
+    cat /var/lib/snapd/system-key | MATCH "build-id: [0-9a-z]+"
```
Comment (zyga) [inline tests/main/snap-system-key/task.yaml]: Can you please add a test that checks that apparmor features is present (regardless of the value)

File: overlord/ifacestate/helpers.go [hunk 5]
```diff
@@ -54,8 +58,10 @@ func (m *InterfaceManager) initialize(extraInterfaces []interfaces.Interface, ex
 	if _, err := m.reloadConnections(""); err != nil {
 		return err
 	}
-	if err := m.regenerateAllSecurityProfiles(); err != nil {
-		return err
+	if m.profilesNeedRegeneration() {
```
Comment (stolowski) [inline overlord/ifacestate/helpers.go]: :heart:

File: tests/main/snap-system-key/task.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    buf="$(stat /var/lib/snapd/system-key)"
+    system_key_content="$(cat /var/lib/snapd/system-key)"
+
+    echo "Ensure that the system-key is not rewritten if system-key is unchanged"
+    restart_snapd
+    buf2="$(stat /var/lib/snapd/system-key)"
+    if [ "$buf" != "$buf2" ]; then
```
Comment (stolowski) [inline tests/main/snap-system-key/task.yaml]: Nice test!

**Comment by mvo5** [comment]
Test is failing because the fact that nfs is used for /home needs to be part of the system key. We should probably also include the kernel version. This probably needs to be its own PR.

**Comment by zyga** [comment]
@mvo5 I'm continuing this in https://github.com/snapcore/snapd/pull/4642
